server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java

```diff
@@ -0,0 +1,214 @@
+        this.joinHelper = new JoinHelper(settings, allocationService, masterService, transportService, this::getCurrentTerm,
+            this::handleJoinRequest);
+        this.persistedStateSupplier = persistedStateSupplier;
+        this.lastKnownLeader = Optional.empty();
+        this.lastJoin = Optional.empty();
+    }
+
+    private Optional<Join> ensureTermAtLeast(DiscoveryNode sourceNode, long targetTerm) {
+        assert Thread.holdsLock(mutex) : "Coordinator mutex not held";
+        if (getCurrentTerm() < targetTerm) {
+            return Optional.of(joinLeaderInTerm(new StartJoinRequest(sourceNode, targetTerm)));
+        }
+        return Optional.empty();
+    }
+
+    private Join joinLeaderInTerm(StartJoinRequest startJoinRequest) {
+        assert Thread.holdsLock(mutex) : "Coordinator mutex not held";
+        logger.debug("joinLeaderInTerm: from [{}] with term {}", startJoinRequest.getSourceNode(), startJoinRequest.getTerm());
+        Join join = coordinationState.get().handleStartJoin(startJoinRequest);
+        lastJoin = Optional.of(join);
+        if (mode != Mode.CANDIDATE) {
+            becomeCandidate("joinLeaderInTerm");
+        }
+        return join;
+    }
+
+    public void handleJoinRequest(JoinRequest joinRequest, JoinCallback joinCallback) {
+        assert Thread.holdsLock(mutex) == false;
+        transportService.connectToNode(joinRequest.getSourceNode());
+
+        synchronized (mutex) {
+            handleJoinRequestUnderLock(joinRequest, joinCallback);
+        }
+    }
+
+    private void handleJoinRequestUnderLock(JoinRequest joinRequest, JoinCallback joinCallback) {
+        assert Thread.holdsLock(mutex) : "Coordinator mutex not held";
+        logger.trace("handleJoinRequestUnderLock: as {}, handling {}", mode, joinRequest);
+
+        final CoordinationState coordState = coordinationState.get();
+        final boolean prevElectionWon = coordState.electionWon();
+
+        if (joinRequest.getOptionalJoin().isPresent()) {
+            final Join join = joinRequest.getOptionalJoin().get();
+            // if someone thinks we should be master, let's add our vote and try to become one
+            // note that the following line should never throw an exception
+            ensureTermAtLeast(getLocalNode(), join.getTerm()).ifPresent(coordState::handleJoin);
+
+            // if we have already won the election, then the actual join does not matter for election purposes
+            if (coordState.electionWon()) {
+                // add join on a best-effort basis
+                try {
+                    coordState.handleJoin(join);
+                } catch (CoordinationStateRejectedException e) {
+                    logger.trace("failed to add join, ignoring", e);
+                }
+            } else {
+                coordState.handleJoin(join); // this might fail and bubble up the exception
+            }
+        }
+
+        if (prevElectionWon == false && coordState.electionWon()) {
```

Comment:
This can't happen as a `LEADER` but can it happen as a `FOLLOWER`? I think it can't, because we must have bumped our term too. Therefore, can this be reorganised as a switch on `mode`?

Reply:
great observation. The scenario I had in mind while writing these conditions was as follows: Assume you want to do a leader handoff. You send a startJoin to all nodes, telling them to join the new prospective leader. Assume the corresponding join from one node arrives on the prospective leader (that is still a follower) before that one has received the start join. Handling of this join with higher term will trigger `ensureTermAtLeast`, turn the node into a candidate, then handle the join, and then reach this point here. So yes, it sounds like we could fold this check into the CANDIDATE branch here.

Reply:
I've simplified this in bd242edfea1